### PR TITLE
feat(ot3): fw update machinery.

### PR DESCRIPTION
**Pull request by amitlissack**
# Overview

The python side of the firmware update. 

This adds a class that initiates the FW update and one that downloads the application data.

# Changelog

- Added `WaitableCallback` for `CanMessenger`. Makes waiting on specific messages a bit easier.
- `FirmwareUpdateInitiater` sends a message to a subsystem that it needs to enter FW update mode. It then waits for the bootloader to become present.
- 'FirmwareUpdateDownloader` uses a `HexRecordProcessor` to chunk up the binary into CAN messages.
- Added update_fw script for performing update.

# Review requests

The goal of this PR is to empower using, exercising, and testing the bootloader. 

There are some holes:
- We bail on all errors
- There's no update progress

# Risk assessment

None

### CODE REVIEW

File: hardware/opentrons_hardware/drivers/can_bus/can_messenger.py [hunk 1]
```diff
@@ -114,3 +116,47 @@ async def _read_task(self) -> None:
...
+
+    def __aiter__(self) -> WaitableCallback:
+        """Enter iterator."""
+        return self
+
+    async def __anext__(self) -> Tuple[MessageDefinition, ArbitrationId]:
+        """Async next."""
+        return await self.read()
+
+    async def read(self) -> Tuple[MessageDefinition, ArbitrationId]:
```
Comment (sfoster1) [inline hardware/opentrons_hardware/drivers/can_bus/can_messenger.py:160]: unless i'm missing something, this just returns the next incoming message, right? should it only return the message if it's actually the one you're looking for, or one of a set that you're looking for?

Comment (sfoster1) [reply hardware/opentrons_hardware/drivers/can_bus/can_messenger.py:160]: I see you've got that implemented in the call site, but I feel like it's a better fit for this callback here.

Comment (sfoster1) [reply hardware/opentrons_hardware/drivers/can_bus/can_messenger.py:160]: Could take a `match` predicate lambda or something

**Review by sfoster1 — APPROVED** [review]
Looks good, though I think we should push down the response matching into the waiter itself rather than having the caller do it. I'd expect it to work that way from the name of the waiter.

File: hardware/opentrons_hardware/drivers/can_bus/can_messenger.py (hunk 1 above)
Comment (amitlissack) [reply hardware/opentrons_hardware/drivers/can_bus/can_messenger.py:160]: Agree. I was thinking that too, but I'd like to push it to a future PR. 

The slight nuisance is that the waiter would still have to do an isinstance check (or cast) on the message due to its being a Union. Or mypy will cry. 

I'd like to move ahead to the C part and tackle this in the future if you're cool with that.

Comment (sfoster1) [reply hardware/opentrons_hardware/drivers/can_bus/can_messenger.py:160]: Sounds good!

**Pull request merged**